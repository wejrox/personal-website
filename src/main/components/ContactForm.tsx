import * as React from "react";
import { Button, Col, Form, FormFeedback, FormGroup, Input, Label, Row } from "reactstrap";

/**
 * @property {string} firstName  (Optional) First name of the person submitting the form.
 * @property {string} lastName   (Optional) Last name/Surname of the person submitting the form.
 * @property {string} email      EMail address of the person submitting the form.
 * @property {string} message    Message to be sent to the database.
 */
interface IContactFormState {
    firstName: string;
    lastName: string;
    email: string;
    message: string;
}

/**
 * A form which contains a first and last name field, alongside an email and message field which will submit a message
 * to the database upon valid submission.
 * The email and message fields are required.
 */
class ContactForm extends React.Component<{}, IContactFormState> {
    constructor(props: {}) {
        super(props);
        this.state = {firstName: "", lastName: "", email: "", message: ""};

        this.handleSubmit = this.handleSubmit.bind(this);
    }

    public render() {
        return (
            <Form onSubmit={this.handleSubmit}>
                <Row>
                    <Col md={6}>
                        <FormGroup>
                            <Label for={"firstName"}>First name</Label>
                            <Input type={"text"} name={"first name"} id={"firstName"}
                                   onChange={(e) => this.setState({firstName: e.target.value})}/>
                        </FormGroup>
                    </Col>
                    <Col md={6}>
                        <FormGroup>
                            <Label for={"lastName"}>Surname</Label>
                            <Input type={"text"} name={"last name"} id={"lastName"}
                                   onChange={(e) => this.setState({lastName: e.target.value})}/>
                        </FormGroup>
                    </Col>
                </Row>
                <FormGroup>
                    <Label for={"emailAddress"}>E-mail address</Label>
                    <Input required type={"email"} name={"email"} id={"emailAddress"}
                           onChange={(e) => this.setState({email: e.target.value})}/>
                </FormGroup>
                <FormGroup>
                    <Label for={"message"}>Message</Label>
                    <Input required type={"textarea"} name={"message"} id={"message"}
                           onChange={(e) => this.setState({message: e.target.value})}/>
                    <FormFeedback hidden={false}>Don't forget to input a message for me!</FormFeedback>
                </FormGroup>
                <Button type={"submit"} id={"submit"}>Send</Button>
            </Form>
        );
    }

    /**
     * TODO: Database connection and submission.
     * Handles the submission of the contact form.
     * Saves the submission into the database for later viewing.
     *
     * @param {React.FormEvent<HTMLFormElement>} event  The form event that triggered the function.
     */
    public handleSubmit(event: React.FormEvent<HTMLFormElement>) {

        // Prevent page refresh.
        event.preventDefault();
        console.log(this.state.message);
    }
}

export default ContactForm;
