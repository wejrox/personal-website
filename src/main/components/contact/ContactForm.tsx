import * as React from "react";
import { Button, Col, Form, FormGroup, Input, Label, Row } from "reactstrap";
import { observable } from "mobx";
import { observer } from "mobx-react";

/**
 * A form which contains a first and last name field, alongside an email and message field which will submit a message
 * to the database upon valid submission.
 * The email and message fields are required.
 */
@observer
class ContactForm extends React.Component {

    // (Optional) First name of the person submitting the form.
    @observable private firstName: string = "";

    // (Optional) Last name/Surname of the person submitting the form.
    @observable private lastName: string = "";

    // EMail address of the person submitting the form.
    @observable private email: string = "";

    // Message to be sent to the database.
    @observable private message: string = "";

    // Whether or not the user has successfully submitted.
    @observable private submitted: boolean = false;

    constructor(props: {}) {
        super(props);

        this.handleSubmit = this.handleSubmit.bind(this);
    }

    public render() {
        return (
            <Form onSubmit={this.handleSubmit}>
                <Row>
                    <Col md={6}>
                        <FormGroup>
                            <Label for={"firstName"}>First name</Label>
                            <Input type={"text"} name={"first name"} id={"firstName"}
                                   onChange={(e) => this.firstName = e.target.value}/>
                        </FormGroup>
                    </Col>
                    <Col md={6}>
                        <FormGroup>
                            <Label for={"lastName"}>Surname</Label>
                            <Input type={"text"} name={"last name"} id={"lastName"}
                                   onChange={(e) => this.lastName = e.target.value}/>
                        </FormGroup>
                    </Col>
                </Row>
                <FormGroup>
                    <Label for={"emailAddress"}>E-mail address</Label>
                    <Input required type={"email"} name={"email"} id={"emailAddress"}
                           onChange={(e) => this.email = e.target.value}/>
                </FormGroup>
                <FormGroup>
                    <Label for={"message"}>Message</Label>
                    <Input required type={"textarea"} name={"message"} id={"message"}
                           onChange={(e) => this.message = e.target.value}/>
                </FormGroup>
                {this.submitted &&
                <div className={"d-flex pb-3 success-message justify-content-center"}>
                    Message sent.
                </div>
                }
                <Button type={"submit"} id={"submit"} className={"btn-block"}>Send</Button>
            </Form>
        );
    }

    /**
     * TODO: Database connection and submission.
     * Handles the submission of the contact form.
     * Saves the submission into the database for later viewing.
     *
     * @param {React.FormEvent<HTMLFormElement>} event  The form event that triggered the function.
     */
    public handleSubmit(event: React.FormEvent<HTMLFormElement>) {

        // Prevent page refresh.
        event.preventDefault();
        this.submitted = true;
        console.log(this.message);
    }
}

export default ContactForm;
