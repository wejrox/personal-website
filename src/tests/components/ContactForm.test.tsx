import React from "react";
import * as Enzyme from "enzyme";
import ContactForm from "../../main/components/contact/ContactForm";

describe("ContactForm Component", () => {
    const wrapper = Enzyme.shallow(<ContactForm/>);

    it("renders without crashing", () => {
        expect(wrapper).toMatchSnapshot();
    });

    describe("State management", () => {
        const dummyNewValue: string = "Something";

        it("Sets the first name correctly", () => {
            wrapper.find("#firstName").simulate("change", {target: {value: dummyNewValue}});

            expect((wrapper.instance() as ContactForm).firstName).toEqual(dummyNewValue);
        });

        it("Sets the last name correctly", () => {
            wrapper.find("#lastName").simulate("change", {target: {value: dummyNewValue}});

            expect((wrapper.instance() as ContactForm).lastName).toEqual(dummyNewValue);
        });

        it("Sets the email address correctly", () => {
            wrapper.find("#emailAddress").simulate("change", {target: {value: dummyNewValue}});

            expect((wrapper.instance() as ContactForm).email).toEqual(dummyNewValue);
        });

        it("Sets the message correctly", () => {
            wrapper.find("#message").simulate("change", {target: {value: dummyNewValue}});

            expect((wrapper.instance() as ContactForm).message).toEqual(dummyNewValue);
        });
    });

    // Validation is performed by an external library so it is not being tested.
    it("Should handle submission correctly", () => {
        const formEvent = {
            preventDefault: () => {
                return;
            },
        } as React.FormEvent<HTMLFormElement>;

        // Currently submission does nothing useful, so there is nothing to test.
        (wrapper.instance() as ContactForm).handleSubmit(formEvent);
    });
});
