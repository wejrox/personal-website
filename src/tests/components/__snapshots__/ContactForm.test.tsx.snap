// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactForm Component renders without crashing 1`] = `ShallowWrapper {}`;
